x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,432 @@
+		VisibilityTimeout: &visibilityTimeout,
+		ReceiptHandle:     receiptHandle,
+	})
+	_, err := req.Send(p.context)
+	return err
+}
+
+func getRegionFromQueueURL(queueURL string) (string, error) {
+	// get region from queueURL
+	// Example: https://sqs.us-east-1.amazonaws.com/627959692251/test-s3-logs
+	queueURLSplit := strings.Split(queueURL, ".")
+	if queueURLSplit[0] == "https://sqs" && queueURLSplit[2] == "amazonaws" {
+		return queueURLSplit[1], nil
+	}
+	return "", errors.New("queueURL is not in format: https://sqs.{REGION_ENDPOINT}.amazonaws.com/{ACCOUNT_NUMBER}/{QUEUE_NAME}")
+}
+
+// handle message
+func handleSQSMessage(m sqs.Message) ([]s3Info, error) {
+	msg := sqsMessage{}
+	err := json.Unmarshal([]byte(*m.Body), &msg)
+	if err != nil {
+		return nil, errors.Wrap(err, "json unmarshal sqs message body failed")
+	}
+
+	var s3Infos []s3Info
+	for _, record := range msg.Records {
+		if record.EventSource == "aws:s3" && record.EventName == "ObjectCreated:Put" {
+			s3Infos = append(s3Infos, s3Info{
+				region: record.AwsRegion,
+				name:   record.S3.bucket.Name,
+				key:    record.S3.object.Key,
+				arn:    record.S3.bucket.Arn,
+			})
+		}
+	}
+	return s3Infos, nil
+}
+
+func (p *Input) handleS3Objects(svc s3iface.ClientAPI, s3Infos []s3Info) error {
+	for _, s3Info := range s3Infos {
+		objectHash := s3ObjectHash(s3Info)
+
+		// read from s3 object
+		reader, err := p.bufferedIORead(svc, s3Info)
+		if err != nil {
+			return errors.Wrap(err, "bufferedIORead failed")
+		}
+
+		offset := 0
+		for {
+			log, err := reader.ReadString('\n')
+			if log == "" {
+				break
+			}
+
+			if err != nil {
+				if err == io.EOF {
+					// create event for last line
+					offset += len([]byte(log))
+					err = p.forwardEvent(createEvent(log, offset, s3Info, objectHash))
+					if err != nil {
+						return errors.Wrapf(err, "forwardEvent failed for %v", s3Info.key)
+					}
+					return nil
+				}
+				return errors.Wrapf(err, "ReadString failed for %v", s3Info.key)
+			}
+
+			// create event per log line
+			offset += len([]byte(log))
+			err = p.forwardEvent(createEvent(log, offset, s3Info, objectHash))
+			if err != nil {
+				return errors.Wrapf(err, "forwardEvent failed for %v", s3Info.key)
+			}
+		}
+	}
+	return nil
+}
+
+func (p *Input) bufferedIORead(svc s3iface.ClientAPI, s3Info s3Info) (*bufio.Reader, error) {
+	s3GetObjectInput := &s3.GetObjectInput{
+		Bucket: awssdk.String(s3Info.name),
+		Key:    awssdk.String(s3Info.key),
+	}
+	req := svc.GetObjectRequest(s3GetObjectInput)
+
+	resp, err := req.Send(p.context)
+	if err != nil {
+		return nil, errors.Wrapf(err, "s3 get object request failed %v", s3Info.key)
+	}
+
+	return bufio.NewReader(resp.Body), nil
+}
+
+func (p *Input) forwardEvent(event *beat.Event) error {
+	forwarder := harvester.NewForwarder(p.outlet)
+	d := &util.Data{Event: *event}
+	err := forwarder.Send(d)
+	if err != nil {
+		return errors.Wrap(err, "forwarder send failed")
+	}
+	return nil
+}
```

Comment:
harvester.NewForwarder just wraps p.outlet, changing the return type from bool to error. We don't really need the forwarder (and should remove it) from our code base.

Reply:
Yeah, so is removing harvester.NewForwarder the plan for the future? I will remove it in s3 input code and use `ok := p.outlet.OnEvent(event)` directly here. Thanks!

Reply:
> Yeah, so is removing harvester.NewForwarder the plan for the future?

Yes, that's the plan. Short term to remove Forwarder, long term to remove Outlet.

Reply:
Good to know, thanks! Forwarder is removed from s3 input 👍